### PR TITLE
Stabilize result object

### LINKED ISSUE #17: memoize useAsyncCallback ?
Hi,

useAsyncCallback returns a new object (and a new execute function) at each call even if the callback passed does not change. It causes re-renders.

### PR DESCRIPTION
**Pull request by fabiosantoscode**
This makes it possible to use the return value from the hooks in this library
as dependencies for other hooks.

Closes #17

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -465,12 +479,17 @@ export const useAsyncFetchMore = <R, Args extends any[]>({
...
+
+  return useMemo(
+    () => ({
+      canFetchMore,
+      loading: fetchMoreAsync.loading,
+      status: fetchMoreAsync.status,
+      fetchMore: fetchMoreAsync.execute,
+      isEnd,
+    }),
+    [canFetchMore, fetchMoreAsync, isEnd]
```
Comment (fabiosantoscode) [inline src/index.ts:493]: `fetchMoreAsync` is memoized so no need to check for its individual methods.

File: src/index.ts [hunk 2]
```diff
@@ -287,15 +288,28 @@ const useAsyncInternal = <R = UnknownResult, Args extends any[] = UnknownArgs>(
...
+    [
+      AsyncState.value,
+      AsyncState.set,
+      AsyncState.merge,
+      AsyncState.reset,
+      executeAsyncOperation,
+      currentPromise,
+      currentParams,
+    ]
+  );
```
Comment (fabiosantoscode) [inline src/index.ts:312]: I noticed that `AsyncState.value` is never mutated, so the memo dependency is just `.value` itself.

**Comment by fabiosantoscode** [comment]
In #17 there's talk of memoizing the `execute` callback itself but I'm not confident I understand the codebase enough to do this.

**Comment by slorber** [comment]
Hey @fabiosantoscode , thanks for this PR

Unfortunately that does not look like something I'd like to merge. I don't think it's very important to guarantee to the user that the return value is memoized, as it does not look to me a good practice to use such big object as a dependency in the first place. You'd rather add to the dependency array individual properties you actually need in the effect (for example `[asyncResult.status, asyncResult.execute]`.


To me the current problem this library has is:

- The execute function is declared inline, this is the first thing we should focus on memoizing: https://github.com/slorber/react-async-hook/blob/168a6448723fa746f4890a28233ac3905d5ef62d/src/index.ts#L247

- The api should probably return something like `[state, api]` instead of a big object. This is a breaking change but this would allow to ensure that at least the api part is always the same object, allowing you to use [api] as dependency.

These are changes I'm planning to make, v2 should have a little bit of breaking changes but that seems easy to migrate.

**Comment by fabiosantoscode** [comment]
You're completely right that execute should be more stable.

However I'm not sure how to fix this, so it's probably best to close.

**Comment by slorber** [comment]
ok, I'm going to handle everything myself when I have some time :)

**Comment by henrymoulton** [comment]
@slorber just checking here - pretty sure I ran into this issue

**Comment by slorber** [comment]
Still an issue yes and I don't think it's worth stabilizing the result object, because it will change anyway.

You'd better use each returned attribute independently.

I just made the "execute()" callback stable so you can use it in deps arrays, other callbacks were already stable

In the future, I'd like to move to `[state, api]` where `api` remains stable and could be used in deps array as a whole

**Pull request closed without merging**